core/node/state_keeper/src/keeper.rs

```diff
@@ -65,48 +67,58 @@ impl Error {
 /// a sequence of executed L2 blocks and batches.
 #[derive(Debug)]
 pub struct ZkSyncStateKeeper {
-    stop_receiver: watch::Receiver<bool>,
     io: Box<dyn StateKeeperIO>,
     output_handler: OutputHandler,
     batch_executor: Box<dyn BatchExecutorFactory<OwnedStorage>>,
     sealer: Arc<dyn ConditionalSealer>,
     storage_factory: Arc<dyn ReadStorageFactory>,
+    health_updater: HealthUpdater,
 }
 
 impl ZkSyncStateKeeper {
     pub fn new(
-        stop_receiver: watch::Receiver<bool>,
         sequencer: Box<dyn StateKeeperIO>,
         batch_executor: Box<dyn BatchExecutorFactory<OwnedStorage>>,
         output_handler: OutputHandler,
         sealer: Arc<dyn ConditionalSealer>,
         storage_factory: Arc<dyn ReadStorageFactory>,
     ) -> Self {
         Self {
-            stop_receiver,
             io: sequencer,
             batch_executor,
             output_handler,
             sealer,
             storage_factory,
+            health_updater: ReactiveHealthCheck::new("state_keeper").1,
         }
     }
 
-    pub async fn run(mut self) -> anyhow::Result<()> {
-        match self.run_inner().await {
+    pub async fn run(mut self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        match self.run_inner(stop_receiver).await {
             Ok(_) => unreachable!(),
-            Err(Error::Fatal(err)) => Err(err).context("state_keeper failed"),
+            Err(Error::Fatal(err)) => {
+                self.health_updater
+                    .update(Health::from(HealthStatus::ShuttingDown));
```

Comment:
Nit: It's really redundant to set the `ShuttingDown` status just before the `HealthUpdater` is dropped, since it'll immediately be overwritten as `ShutDown`. `ShuttingDown` makes sense to set before non-trivial shut down logic.

Reply:
Addressed in https://github.com/matter-labs/zksync-era/pull/3193/commits/53cca2fb8d510e0cea578a02c73a987bb48909be